### PR TITLE
Add 4dmask support for attention cuda kernel 

### PR DESCRIPTION
**Pull request by wangyems**
**Description**: Describe your changes.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: onnxruntime/contrib_ops/cpu/bert/attention.cc [hunk 1]
```diff
@@ -146,8 +146,12 @@ Status AttentionBase::CheckInputs(const TensorShape& input_shape,
       if (static_cast<int>(mask_dims[0]) != batch_size || mask_dims[1] != sequence_length || static_cast<int>(mask_dims[2]) != past_sequence_length + sequence_length) {
         return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Inputs 'mask_index' with 3D data shall have shape batch_size x sequence_length x (past_sequence_length + sequence_length)");
       }
+    } else if (mask_dims.size() == 4) {
+      if (static_cast<int>(mask_dims[0]) != batch_size || mask_dims[1] != 1 || mask_dims[2] != mask_dims[3] || mask_dims[2] < past_sequence_length + sequence_length) {
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/bert/attention.cc:150]: For mask_dims[2] < past_sequence_length + sequence_length, why uses < instead of !=?

File: onnxruntime/contrib_ops/cuda/bert/attention_softmax.h [hunk 2]
```diff
@@ -180,7 +181,16 @@ __device__ inline void SoftmaxWithRawMaskSmall(const int all_sequence_length,
   if (threadIdx.x < all_sequence_length) {
     const int batch_index = blockIdx.y;
     const int sequence_index = blockIdx.x % sequence_length;
-    const int mask_offset = (mask_dimension == 2) ? batch_index * all_sequence_length + threadIdx.x : batch_index * sequence_length * all_sequence_length + sequence_index * all_sequence_length + threadIdx.x;
+    int mask_offset = 0;
+    if (mask_dimension == 2) {
+      mask_offset = batch_index * all_sequence_length + threadIdx.x;
+    } else if (mask_dimension == 3) {
+      mask_offset = batch_index * sequence_length * all_sequence_length + sequence_index * all_sequence_length + threadIdx.x;
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/bert/attention_softmax.h:188]: Use the following might be faster (4 operations instead of 5):
```
(batch_index * sequence_length + sequence_index) * all_sequence_length + threadIdex.x
```

For 4D, it could have similar change.

File: onnxruntime/contrib_ops/cuda/bert/attention_softmax.h [hunk 3]
```diff
@@ -180,7 +181,16 @@ __device__ inline void SoftmaxWithRawMaskSmall(const int all_sequence_length,
...
+    } else if (mask_dimension == 4){
+      // Megatron code:
+      // ltor_mask = ltor_mask[..., (attention_scores.size(3)-hidden_states.size(1)):attention_scores.size(3), :attention_scores.size(3)]
+      mask_offset = batch_index * max_sequence_length * max_sequence_length + (all_sequence_length - sequence_length) * max_sequence_length + sequence_index * max_sequence_length + threadIdx.x;
+    }
 
     const int& mask = attention_mask[mask_offset];
     float mask_value = mask > 0 ? 0.0f : -10000.0f;
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/bert/attention_softmax.h:196]: Need also consider the effect of is_unidirectional and 4d mask as in next few lines.  For example, change 
```
 if (is_unidirectional) {
```
to 
```
 if (is_unidirectional && mask_dimension != 4) {
```

File: onnxruntime/test/contrib_ops/attention_op_test.cc [hunk 4]
```diff
@@ -1363,6 +1370,50 @@ TEST(AttentionTest, AttentionDummyMask2D) {
...
+      0, 1, 0, 0,
+      0, 1, 0, 0,
+      0, 0, 0, 0,
+      0, 0, 0, 0};
+
+  std::vector<float> output_data = {
+      8.69f, -0.13f, 4.25f, 5.65f,
+      8.69f, -0.13f, 4.25f, 5.65f};
+
+  bool use_float16 = true;
```
Comment (tianleiwu) [inline onnxruntime/test/contrib_ops/attention_op_test.cc:1403]: I suggest add a test case for fp32 at least. It is because fp16 test might not run in pipeline since the gpu is old.

File: onnxruntime/test/contrib_ops/attention_op_test.cc [hunk 5]
```diff
@@ -1363,6 +1370,50 @@ TEST(AttentionTest, AttentionDummyMask2D) {
...
+  std::vector<float> weight_data = {
+      0.1f, -0.2f, 0.3f, 1.0f, 1.1f, 0.3f, 0.5f, 0.2f, 0.3f, -0.6f, 1.5f, 2.0f,
+      0.5f, 0.1f, 0.4f, 1.6f, 1.0f, 2.0f, 0.4f, 0.8f, 0.9f, 0.1f, -1.3f, 0.7f,
+      0.3f, 0.2f, 4.0f, 2.2f, 1.6f, 1.1f, 0.7f, 0.2f, 0.4f, 1.0f, 1.2f, 0.5f,
+      0.2f, 0.1f, 0.4f, 1.6f, 2.4f, 3.3f, 2.1f, 4.2f, 8.4f, 0.0f, 2.1f, 3.2f};
+
+  std::vector<float> bias_data = {
+      -0.5f, 0.6f, 1.2f, 2.1f, 0.5f, 0.7f, 0.2f, 1.2f, 0.5f, 0.4f, 0.3f, 1.2f};
+
+  // Test 3D mask BxSxS*
```
Comment (tianleiwu) [inline onnxruntime/test/contrib_ops/attention_op_test.cc:1392]: update comment

**Review by tianleiwu — APPROVED** [review]
